pywbem_mock/_baserepository.py

```diff
@@ -0,0 +1,166 @@
+#
+# (C) Copyright 2020 InovaDevelopment.comn
+#
+# This library is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2.1 of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+#
+# Author: Karl  Schopmeyer <inovadevelopment.com>
+#
+
+
+"""
+Mock support for the WBEMConnection class to allow pywbem users to test the
+pywbem client without requiring a running WBEM server.
+
+For documentation, see mocksupport.rst.
+"""
+from abc import ABCMeta, abstractmethod
+
+import six
+
+from pywbem import CIMClass, CIMQualifierDeclaration, CIMInstance
+
+
+@six.add_metaclass(ABCMeta)
+class BaseObjStore(object):
+
+    def __init__(self, case_insensitive_names, cim_object_type):
+        """
+        """
+        # TODO namespace must point to dynamic object that changes when
+        # namespaces added.
+        self._case_insensitive_names = case_insensitive_names
+        self._cim_object_type = cim_object_type
+
+    @abstractmethod
+    def exists(self, name):
+        """
+        Test if the object defined by name and namespace exists in the object
+        store
+        """
+        pass
+
+    @abstractmethod
+    def get(self, name, copy=True):
+        """
+        Return the object defined by name and namespace if it exists.
+
+        Returns:
+
+        Raises:
+            TODO: What do we return for namespace error.
+            KeyError: Object not in store
+        """
+        pass
+
+    @abstractmethod
+    def create(self, name, object):
+        """
+        """
+        pass
+
+    @abstractmethod
+    def update(self, name, object):
+        """
+        """
+        pass
+
+    @abstractmethod
+    def delete(self, name):
+        """
+        """
+        pass
+
+    def names(self):
+        """
+        Return all of the objects defined in this data store
```

Comment:
Should we make this also return an iterator, instead of materializing it to a list? The caller can always materialize an iterator to a list. If so, name it "iter_names".